app/controllers/projects/tree_filter_controller.rb

```diff
@@ -0,0 +1,20 @@
+# Controller for viewing a repository's file structure
+class Projects::TreeFilterController < Projects::ApplicationController
+  include ExtractsPath
+  include ActionView::Helpers::SanitizeHelper
+
+  before_action :assign_ref_vars
+
+  def show
+    return render_404 unless @repository.commit(@ref)

```

Comment:
Do you need to check whether the current user has permission to this repository?